File: branches/COMET/src/graphics.js
/**
 * graphics.js
 * Graphics Routines for Crunchy, these are all primitive routines accessed
 * by the Python code running on the server.
 * 
 * Changelog:
 * 6 Feb 07: File Created, imports from the old crunchy graphics
 * 21 Feb 07: A few Fixes added, now it all seems to work
 */
 
// get the context for the given uid:
function get_ctx(uid){
    return document.getElementById("canvas_" + uid).getContext('2d');
}

// initialisation, displays the corresponding graphics element and blanks it
function init_graphics(uid){
    //alert(uid);
    //todo: implement width and height
    document.getElementById("canvas_" + uid).style.display = "block";
    clear_rect(uid, 0, 0, 400, 400);
    //alert(uid);
}

function clear_rect(uid, x, y, width, height){
    get_ctx(uid).clearRect(x, y, width, height);
}

function set_line_colour(uid, colour){
    get_ctx(uid).strokeStyle = colour;
}

function set_fill_colour(uid, colour){
    get_ctx(uid).fillStyle = colour;
}

function draw_line(uid, x1, y1, x2, y2){
    get_ctx(uid).beginPath();
    get_ctx(uid).moveTo(x1, y1);
    get_ctx(uid).lineTo(x2, y2);
    get_ctx(uid).stroke();
}

function circle_stroke(uid, x, y, r){
    get_ctx(uid).beginPath();
    get_ctx(uid).arc(x, y, r, 0, Math.PI*2, true);    //hopefully this works :)
    get_ctx(uid).stroke();
}

function circle_fill(uid, x, y, r){
    get_ctx(uid).beginPath();
    get_ctx(uid).arc(x, y, r, 0, Math.PI*2, true); 
    get_ctx(uid).fill();
}

function rectangle_stroke(uid, x, y, width, height){
    get_ctx(uid).strokeRect(x, y, width, height);
}

function rectangle_fill(uid, x, y, width, height){
    get_ctx(uid).fillRect(x, y, width, height);
}

function triangle_stroke(uid, x1, y1, x2, y2, x3, y3){
    get_ctx(uid).beginPath();
    get_ctx(uid).moveTo(x1, y1);
    get_ctx(uid).lineTo(x2, y2);
    get_ctx(uid).lineTo(x3, y3);
    get_ctx(uid).closePath();
    get_ctx(uid).stroke();
}

function triangle_fill(uid, x1, y1, x2, y2, x3, y3){
    get_ctx(uid).beginPath();
    get_ctx(uid).moveTo(x1, y1);
    get_ctx(uid).lineTo(x2, y2);
    get_ctx(uid).lineTo(x3, y3);
    get_ctx(uid).closePath();
    get_ctx(uid).fill();
}

File: branches/johann/plugins/graphics.py
"""
graphics.py
"""

import threading
import re

import CrunchyPlugin as __cp

def register():
    # no callbacks to register or initialisation needed
    pass

def init_graphics(width=400, height=400, border_color='red'):
    uid = __cp.get_uid()
    __cp.exec_js(__cp.get_pageid(), """document.getElementById("canvas_%s").width=%d;
                    document.getElementById("canvas_%s").height=%d;
                    document.getElementById("canvas_%s").style.display = "block";
                    document.getElementById("canvas_%s").getContext('2d').clearRect(0, 0, %d, %d);
                    """ % (uid, width, uid, height, uid, uid, width, height))
    set_line_colour('%s'%border_color)
    rectangle((0, 0), width, height)
    set_line_colour('black')

def set_line_colour(col):
    uid = __cp.get_uid()
    if not validate_colour(col):
        col = "DeepPink" # make it stand out for now
    __cp.exec_js(__cp.get_pageid(), """document.getElementById("canvas_%s").getContext('2d').strokeStyle = %r;""" % (uid, col))
set_line_color = set_line_colour # American spelling == British/Canadian spelling

def set_fill_colour(col):
    uid = __cp.get_uid()
    if not validate_colour(col):
        col = "DeepPink" # make it stand out for now
    __cp.exec_js(__cp.get_pageid(), """document.getElementById("canvas_%s").getContext('2d').fillStyle = %r;""" % (uid, col))
set_fill_color = set_fill_colour

def line((x1, y1), (x2, y2)):
    uid = __cp.get_uid()
    __cp.exec_js(__cp.get_pageid(), """document.getElementById("canvas_%s").getContext('2d').beginPath();
                             document.getElementById("canvas_%s").getContext('2d').moveTo(%s, %s);
                             document.getElementById("canvas_%s").getContext('2d').lineTo(%s, %s);
                             document.getElementById("canvas_%s").getContext('2d').stroke();""" % (uid, uid, x1, y1, uid, x2, y2, uid))

def circle((x, y), r):
    uid = __cp.get_uid()
    __cp.exec_js(__cp.get_pageid(), """document.getElementById("canvas_%s").getContext('2d').beginPath();
                             document.getElementById("canvas_%s").getContext('2d').arc(%s, %s, %s, 0, Math.PI*2, true);
                             document.getElementById("canvas_%s").getContext('2d').stroke();""" % (uid, uid, x, y, r, uid))

def filled_circle((x, y), r):
    uid = __cp.get_uid()
    __cp.exec_js(__cp.get_pageid(), """document.getElementById("canvas_%s").getContext('2d').beginPath();
                             document.getElementById("canvas_%s").getContext('2d').arc(%s, %s, %s, 0, Math.PI*2, true);
                             document.getElementById("canvas_%s").getContext('2d').fill();""" % (uid, uid, x, y, r, uid))

def rectangle((x1, y1), w, h):
    uid = __cp.get_uid()
    __cp.exec_js(__cp.get_pageid(), """document.getElementById("canvas_%s").getContext('2d').strokeRect(%s, %s, %s, %s);""" % (uid, x1, y1, w, h))

def filled_rectangle((x1, y1), w, h):
    uid = __cp.get_uid()
    __cp.exec_js(__cp.get_pageid(), """document.getElementById("canvas_%s").getContext('2d').fillRect(%s, %s, %s, %s);""" % (uid, x1, y1, w, h))

def triangle((x1, y1), (x2, y2), (x3, y3)):
    uid = __cp.get_uid()
    __cp.exec_js(__cp.get_pageid(), """document.getElementById("canvas_%s").getContext('2d').beginPath();
                             document.getElementById("canvas_%s").getContext('2d').moveTo(%s, %s);
                             document.getElementById("canvas_%s").getContext('2d').lineTo(%s, %s);
                             document.getElementById("canvas_%s").getContext('2d').lineTo(%s, %s);
                             document.getElementById("canvas_%s").getContext('2d').closePath();
                             document.getElementById("canvas_%s").getContext('2d').stroke();""" % (uid, uid, x1, y1, uid, x2, y2, uid, x3, y3, uid, uid))

def filled_triangle((x1, y1), (x2, y2), (x3, y3)):
    uid = __cp.get_uid()
    __cp.exec_js(__cp.get_pageid(), """document.getElementById("canvas_%s").getContext('2d').beginPath();
                             document.getElementById("canvas_%s").getContext('2d').moveTo(%s, %s);
                             document.getElementById("canvas_%s").getContext('2d').lineTo(%s, %s);
                             document.getElementById("canvas_%s").getContext('2d').lineTo(%s, %s);
                             document.getElementById("canvas_%s").getContext('2d').closePath();
                             document.getElementById("canvas_%s").getContext('2d').fill();""" % (uid, uid, x1, y1, uid, x2, y2, uid, x3, y3, uid, uid))

def point(x, y):
    uid = __cp.get_uid()
    __cp.exec_js(__cp.get_pageid(), """document.getElementById("canvas_%s").getContext('2d').beginPath();
                             document.getElementById("canvas_%s").getContext('2d').moveTo(%s, %s);
                             document.getElementById("canvas_%s").getContext('2d').lineTo(%s, %s);
                             document.getElementById("canvas_%s").getContext('2d').stroke();""" % (uid, uid, x, y, uid, x+1, y+1, uid))



#--- colour validation
named_colour = re.compile('^[a-zA-Z]*[a-zA-Z]$') # only letters
hex_code = re.compile('^#[a-fA-F0-9]{5,5}[a-fA-F0-9]$') # begin with "#", then 6 hexadecimal values
rgb_pattern = re.compile('^rgb\s*\((.+?),(.+?),(.+?)\)$') #begins with rgb(, etc
rgba_pattern = re.compile('^rgba\s*\((.+?),(.+?),(.+?),(.+?)\)$')

def validate_colour(colour):
    '''verifies that the colour given follows an acceptable pattern'''
    colour = colour.strip()
    if hex_code.match(colour):
        return True
    elif named_colour.match(colour): # ; assume valid as colour name;
        return True           # javascript would treat it as black otherwise.
    elif rgb_pattern.match(colour):
        res = rgb_pattern.search(colour)
        try:
            r = int(res.groups()[0])
            g = int(res.groups()[1])
            b = int(res.groups()[2])
            if r >= 0 and r <= 255 and g >= 0 and g <= 255 and b >= 0 and b <= 255:
                return True
            else:
                return False
                #raise errors.ColourNameError(colour)
        except:
            return False
            #raise errors.ColourNameError(colour)
    elif rgba_pattern.match(colour):
        res = rgba_pattern.search(colour)
        try:
            r = int(res.groups()[0])
            g = int(res.groups()[1])
            b = int(res.groups()[2])
            a = float(res.groups()[3])
            if r >= 0 and r <= 255 and g >= 0 and g <= 255 and b >= 0 and b <= 255 \
                and a >= 0. and a <= 1.0:
                return True
            else:
                return False
                #raise errors.ColourNameError(colour)
        except:
            return False
            #raise errors.ColourNameError(colour)
    return False
    #raise errors.ColourNameError(colour)

